tests/debugging/function/test_store.py

```diff
@@ -0,0 +1,228 @@
+    with FunctionStore() as store:
+        function = FunctionDiscovery.from_module(stuff).by_name(
+            ".".join((stuff.Stuff.__name__, stuff.Stuff.doublydecoratedstuff.__name__))
+        )
+
+        if PY2:
+            assert function.__code__ is stuff.Stuff.doublydecoratedstuff.__func__.__code__
+        else:
+            assert function is stuff.Stuff.doublydecoratedstuff
+
+        arg = mock.Mock()
+        store.wrap(function, gen_wrapper(arg, 42))
+
+        s = stuff.Stuff()
+        s.doublydecoratedstuff()
+
+        arg.assert_called_once_with(42)
+
+        store.restore_all()
+        s.doublydecoratedstuff()
+        arg.assert_called_once_with(42)
+
+
+def test_function_unwrap():
+    with FunctionStore() as store:
+        function = FunctionDiscovery.from_module(stuff).by_name(stuff.modulestuff.__name__)
+        assert function is stuff.modulestuff
+        code = function.__code__
+
+        store.wrap(function, gen_wrapper(None, 42))
+        assert code is not stuff.modulestuff.__code__
+
+        store.unwrap(stuff.modulestuff)
+        assert code is stuff.modulestuff.__code__
+
+
+def test_function_inject_wrap_commutativity():
+    with FunctionStore() as store:
+        code = stuff.modulestuff.__code__
+
+        # Injection
+        lo = min(linenos(stuff.modulestuff))
+        function = FunctionDiscovery.from_module(stuff).at_line(lo)[0]
+        hook = mock.Mock()()
+        store.inject_hook(function, hook, lo, 42)
+
+        # Wrapping
+        function = FunctionDiscovery.from_module(stuff).by_name(stuff.modulestuff.__name__)
+        assert function.__code__ is stuff.modulestuff.__code__
+        store.wrap(function, gen_wrapper(None, 42))
+        assert code is not stuff.modulestuff.__code__
+
+        # Ejection
+        assert stuff.modulestuff.__code__ is not code
+        store.eject_hook(stuff.modulestuff, hook, lo, 42)
+
+        # Unwrapping
+        store.unwrap(stuff.modulestuff)
+
+        assert stuff.modulestuff.__code__ is not code and stuff.modulestuff.__code__ == code
+
+
+def test_function_wrap_inject_commutativity():
+    with FunctionStore() as store:
+        # Wrapping
+        function = FunctionDiscovery.from_module(stuff).by_name(stuff.modulestuff.__name__)
+        assert function is stuff.modulestuff
+        code = function.__code__
+        store.wrap(function, gen_wrapper(None, 42))
+        assert code is not stuff.modulestuff.__code__
+
+        # Injection
+        lo = min(linenos(stuff.modulestuff.__dd_wrapped__))
+        function = FunctionDiscovery.from_module(stuff).at_line(lo)[0]
+        hook = mock.Mock()()
+        store.inject_hook(function, hook, lo, 42)
+
+        # Unwrapping
+        store.unwrap(stuff.modulestuff)
+
+        # Ejection
+        assert stuff.modulestuff.__code__ is not code
+        store.eject_hook(stuff.modulestuff, hook, lo, 42)
+
+        assert stuff.modulestuff.__code__ is not code and stuff.modulestuff.__code__ == code
```

Comment:
@Kyle-Verhoog I am adding multiple wrapping support to the lower-level API in `ddtrace.internal.wrapping` with #3874. Within the remits of `ddtrace.debugging` we only wrap once and use a collection object for handling multiple wrapping. This is so that we have a single wrapping frame (and hence a single additional function call) in the call stack.

Reply:
@P403n1x87 should we still have a test here for multiple wrapping?

Reply:
The function store, as part of the `debugging` submodule, supports single wrapping. That is we expect to wrap once per store instance. The only think I can think of is that perhaps that a second call to `wrap` should maybe unwrap and then rewrap with a new wrapper, but ideally this shouldn't happen. Since this is an internal component, we should have full control over it anyway. WDYT?